### ISSUE TITLE
Figure out how to handle links

### ISSUE THREAD

**Issue by skipjack** [skipjack/react-banner#1]
Need to decide how best to handle links, I see a few routes:

- Make `react-router` a peer dependency and just use `Link` components. My concern initially was that this wouldn't work for non-spa's but because many static site generator's use `react-router` in the background, maybe it's not a problem?
- Allow a custom link prop to be passed to the banner which accepts standardized props. Thus the user can decide what type of link to use.
- Just use normal `a` tags -- the problem with this is it will trigger full reloads even for SPAs.
- Allow browser history to be passed and use it to modify location, else just use normal links.

Still figuring out which would be the best approach...

**Comment by bebraw** [comment]
> Make react-router a peer dependency and just use Link components. My concern initially was that this wouldn't work for non-spa's but because many static site generator's use react-router in the background, maybe it's not a problem?

I would go with an injection pattern. Expose a prop just for that and default to `a`.

> Allow a custom link prop to be passed to the banner which accepts standardized props. Thus the user can decide what type of link to use.

Exactly this. All these kind of aspects can be grouped behind a single prop to keep the API lean.

**Comment by skipjack** [comment]
Working on this now, I'll get a rough draft going and then we can spec out the details (e.g. what props the component takes and how they're named).

**Comment by skipjack** [comment]
@bebraw ok so as of right now I have a rough implementation going... basically I created [two types of links](https://github.com/skipjack/react-banner/tree/master/src/links), one for single page apps and one for normal sites. `banner.jsx` and `banner-sub.jsx` both get everything from each link object passed spread onto them as props. So right now it's not really standardized yet because the link components are just targeted towards the [data](https://github.com/skipjack/react-banner/blob/master/src/site/site-links.json) I'm passing.

Here are a few of the keys that would probably need to be standardized:

- `url` or `href`: The URL so we can determine if a link is active.
- `component`: A custom component so things like icons and dropdowns can be rendered in place of normal links. The props this accepts may have to be standardized a bit too.

I'm going to start documenting props etc but I'll wait on writing the links spec until we've talked more. Let me know your thoughts, I know you've already mentioned a few ideas before.

**Comment by bebraw** [comment]
> url or href: The URL so we can determine if a link is active.

I don't have any strong opinion on this. The current (`url`) is good enough for an initial release.

> component: A custom component so things like icons and dropdowns can be rendered in place of normal links. The props this accepts may have to be standardized a bit too.

Yeah, having an override like that is a cool thing to have. I solved this in Reactabular through a function interface which has been documented. Actually now that I think of it, you could override `title` of the definition like this so you would end up with `title: () => <CustomThinger />`.

> I'm going to start documenting props etc but I'll wait on writing the links spec until we've talked more. Let me know your thoughts, I know you've already mentioned a few ideas before.

A good way to handle this is to generate the site from component propTypes. [react-styleguidist](https://github.com/styleguidist/react-styleguidist) would work, but feel free to pick whatever site generator you are the most comfortable with.

**Comment by skipjack** [comment]
> Actually now that I think of it, you could override `title` of the definition like this so you would end up with `title: () => <CustomThinger />`.

Ahh, that is a good idea I might try that out.

> A good way to handle this is to generate the site from component propTypes. react-styleguidist would work, but feel free to pick whatever site generator you are the most comfortable with.

Yeah, I've played around a little with react-styleguidist but I think I want a bit more control of the site so I might keep going the custom route for now. Doing it this way should allow me to do more in-depth examples and so on. However, I will write all the content in markdown so porting any of this to a react static site generator at some point shouldn't be too bad. 

I guess another option is maybe having a reference page and using react-styleguidist to generate that so it's easy to keep up to date. Do you know of any examples I could take a look where people embed a react-styleguidist generated page within a parent site? All the examples I saw just seemed to be one pagers fully generated by react-styleguidist.

**Comment by bebraw** [comment]
@sapegin Any idea how to use react-styleguidist as a part of another site?

**Comment by sapegin** [comment]
Unfortunately it’s not possible now (maybe a few refatorings and pull requests later but it’s not a priority now). It’s very flexible (you can replace any component, tweak styles, etc.) so you can make a [simple documentation site](https://github.com/styleguidist/react-styleguidist/tree/next/examples/customised) with just Styleguidist. And we’ll have static rendering soon.

**Comment by sapegin** [comment]
You can also redefine HTML template but it may be tricky ;-)

**Comment by skipjack** [comment]
@sapegin thanks for the advice and the pointer to the more customized version. I'm going to continue going the custom route (without react-styleguidist) for now but I may port the docs to your tool or another after the initial few releases. 

@bebraw I'm gonna close this as I think we have a good starting point for the links. We can always revisit it more later.
